### PR TITLE
Recent update in SpeechLM development. 

### PR DESCRIPTION
**Pull request by jctian98**
(1) add more huggingface support, such as huggingface-based model inference code;
(2) revised the collect_stats for my flexible dataset fusion.
(3) add speech enhancement data simulation tool and offline text tokenization tool
(4) revise loss computing
(5) provide recipes for ASR and TTS pre-trained model
(6) add ASR-related evaluation protocols. 
(7) revised inference code and dataloader preprocessor.
(8) revise README to describe the pre-trained models.

May add some minor revisions to this PR, but basically it should be ok to be distributed to collaborators who work on subtasks using the pre-trained models.

For collaborators who want to work with the pre-trained model, please refer to [the resources section of the README file](https://github.com/jctian98/espnet/blob/speechlm_dev/egs2/TEMPLATE/speechlm1/README.md#resources)

### CODE REVIEW

**Comment by ftshijt** [comment]
Since it is a branch-related PR and its dependency to several other collaborators, please let me know if the current shape is ready to merge!

**Comment by jctian98** [comment]
i've checked the process and I think it's ready to merge.
